indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/ParallelIndexSupervisorTask.java

```diff
@@ -360,43 +362,77 @@ SegmentIdWithShardSpec allocateNewSegment(DateTime timestamp) throws IOException
+
     // List locks whenever allocating a new segment because locks might be revoked and no longer valid.
     final Map<Interval, String> versions = toolbox
         .getTaskActionClient()
         .submit(new LockListAction())
         .stream()
+        // FIXME note that the next line is new in this commit --- it's not relevant for fixing #6989,
+        //       it just seems like a bug fix to make the code match its comment
+        .filter(taskLock -> !taskLock.isRevoked())
         .collect(Collectors.toMap(TaskLock::getInterval, TaskLock::getVersion));
 
-    final Optional<Interval> maybeInterval = granularitySpec.bucketInterval(timestamp);
-    if (!maybeInterval.isPresent()) {
-      throw new IAE("Could not find interval for timestamp [%s]", timestamp);
-    }
+    Interval interval;
+    String version;
+    if (bucketIntervals.isPresent()) {
+      // If the granularity spec has explicit intervals, we just need to find the interval (of the segment
+      // granularity); we already tried to lock it at task startup.
+      final Optional<Interval> maybeInterval = granularitySpec.bucketInterval(timestamp);
+      if (!maybeInterval.isPresent()) {
+        throw new IAE("Could not find interval for timestamp [%s]", timestamp);
+      }
+
+      interval = maybeInterval.get();
+      if (!bucketIntervals.get().contains(interval)) {
+        throw new ISE("Unspecified interval[%s] in granularitySpec[%s]", interval, granularitySpec);
+      }
 
-    final Interval interval = maybeInterval.get();
-    if (!bucketIntervals.contains(interval)) {
-      throw new ISE("Unspecified interval[%s] in granularitySpec[%s]", interval, granularitySpec);
+      version = findVersion(versions, interval);
+      if (version == null) {
+        // FIXME I'm preserving the existing error message here, but it seems like the message should
+        // be more like "lock on interval has been revoked"?
+        throw new ISE("Cannot find a version for interval[%s]", interval);
+      }
+    } else {
+      // We don't have explicit intervals. We can use the segment granularity to figure out what
+      // interval we need, but we might not have already locked it.
+      interval = granularitySpec.getSegmentGranularity().bucket(timestamp);
+      version = findVersion(versions, interval);
+      if (version == null) {
+        // We don't have a lock for this interval, so we should lock it now.
+        // FIXME should this be doing a LockAcquireAction instead so it can retry if it fails?
+        //       I think not because you really shouldn't be batch ingesting data that overlaps
+        //       with real time.
+        final TaskLock lock = Preconditions.checkNotNull(
+            toolbox.getTaskActionClient().submit(new LockTryAcquireAction(TaskLockType.EXCLUSIVE, interval)),
+            "Cannot acquire a lock for interval[%s]", interval
+        );
+        if (!interval.equals(lock.getInterval())) {
```

Comment:
I think it's fine to skip this check. The lock is created using the given interval, and so this must not happen.

Reply:
Instead, maybe we want to check this is the first partition of this interval. Probably worth to check `Counters.incrementAndGetInt(partitionNumCountersPerInterval, interval)` returns 1.

Reply:
BTW, `partitionNum` should actually be `Counters.getAndIncrementInt()`. I'm fine with fixing it in this PR if you want.

Reply:
I'm not sure what you mean by the third comment: you mean that the method's name should be changed from Counters.incrementAndGetInt to Counters.getAndIncrementInt?

Reply:
Ah sorry. I mean, we should add a new method which gets first and then increment the counter.

Reply:
Ah, to make it start at 0 rather than 1? OK, done. Note that this was the only caller of `Counters.incrementAndGetInt`: should I delete it, or leave it around for extension compatibility reasons?

Reply:
I think no one is using it. You should be fine to remove it.